### ISSUE TITLE
Support for "validate" helper within rule blocks

### ISSUE THREAD

**Issue by solnic** [dry-rb/dry-validation#679]
Support for executing predicates using `validate` method within `rule` blocks will make it much easier to apply various checks with extra conditions:

```ruby
params do
  optional(:per_page).filled(:integer)
end

rule(:per_page) do
  if key?
    key.failure("must be within 0..20 range") unless validate(gt?: 0, lteq?: 20)
  end
end
```

See the original conversation: https://github.com/dry-rb/dry-validation/issues/678

**Comment by musaffa** [comment]
Will it be possible to call `validate` inside rule block without providing an explicit key failure as shown above? Predicate macros and custom registered macros have key failures defined in them already.

**Comment by flash-gordon** [comment]
smth like `valid?(gt?: 0, lteq?: 20)`. I'm not sure if it should be done in a single PR

**Comment by solnic** [comment]
> Will it be possible to call validate inside rule block without providing an explicit key failure as shown above?

@musaffa yes! I just updated the example but I think @flash-gordon has a point and we should have `valid?` too and I agree it should be in a separate PR as it would be a distinct new feature too.

**Comment by zavan** [comment]
I monkey-patched `Evaluator` to add `validate` as a proof of concept:

```ruby
module Dry
  module Validation
    class Evaluator
      def validate(*args)
        macros = args.each_with_object([]) do |spec, macros|
          case spec
          when Hash
            spec.each do |k, v|
              macros << [k, v.is_a?(Array) ? v : [v]]
            end
          else
            macros << Array(spec)
          end
        end
  
        macros.each do |args|
          macro = macro(*args.flatten(1))
          instance_exec(**macro.extract_block_options(_options.merge(macro: macro)), &macro.block)
        end
      end
    end
  end
end
```

Then you can call `validate(:filled?, included_in?: [[1, 2, 3]], gt?: 18)` within the rule block.

I just took [these methods](https://github.com/dry-rb/dry-validation/blob/664ae8c823a533cc0bccf90f9e7078bed0f8c451/lib/dry/validation/rule.rb#L117-L132) from `Rule` to prepare the macros and [this](https://github.com/dry-rb/dry-validation/blob/664ae8c823a533cc0bccf90f9e7078bed0f8c451/lib/dry/validation/evaluator.rb#L77-L80) from `Evaluator` to run the macros and merged them.

I want to refactor it, write tests and docs and open a PR, but I wanted to clear this out with y'all first: I think in `Rule` we just [assign `@macros` as it's passed (`@macros = macros`)](https://github.com/dry-rb/dry-validation/blob/664ae8c823a533cc0bccf90f9e7078bed0f8c451/lib/dry/validation/rule.rb#L54) to `rule(:name).validate(included_in?: [[1, 2, 3]], gt?: 18)` and move all the [macro parsing](https://github.com/dry-rb/dry-validation/blob/664ae8c823a533cc0bccf90f9e7078bed0f8c451/lib/dry/validation/evaluator.rb#L77-L80) to `Evaluator`. Sounds good?

**Comment by zavan** [comment]
I forked and created a branch if you want to take a look: https://github.com/dry-rb/dry-validation/compare/main...zavan:dry-validation:macros-within-rule

No tests yet and it breaks the existing ones relying on `Contract` doubles, as contract is expected to respond to `#macro`.

**Comment by solnic** [comment]
It would be great to have this. If you could open a PR with fixed tests that would be awesome.
